File: cpp_examples/common/tracker.cpp
#include "tracker.h"

#include <cmath>
#include <limits>
#include <set>

CentroidTracker::CentroidTracker(int maxMissed, float maxDistance)
    : m_maxMissed(maxMissed), m_maxDistance(maxDistance) {}

namespace {

float cosineSim(const std::vector<float>& a, const std::vector<float>& b) {
    if (a.size() != b.size() || a.empty()) return 0.0f;
    float dot = 0.0f;
    for (size_t i = 0; i < a.size(); ++i) dot += a[i] * b[i];
    return dot;   // assumes both are L2-normalised
}

void emaUpdate(std::vector<float>& dst, const std::vector<float>& src) {
    if (src.empty()) return;
    if (dst.size() != src.size()) { dst = src; return; }
    float sumSq = 0.0f;
    for (size_t i = 0; i < dst.size(); ++i) {
        dst[i] = 0.9f * dst[i] + 0.1f * src[i];
        sumSq += dst[i] * dst[i];
    }
    const float norm = std::sqrt(sumSq);
    if (norm > 1e-6f) for (auto& v : dst) v /= norm;
}

} // namespace

void CentroidTracker::reset() {
    m_tracks.clear();
    m_nextId = 0;
}

std::map<int, int> CentroidTracker::update(
    const std::vector<cv::Point2f>& centroids)
{
    // Step 1: age every existing track. Matched tracks below will have
    // missedFrames reset back to 0.
    for (auto& kv : m_tracks) kv.second.missedFrames++;

    std::map<int, int> assignments;     // input index -> track id
    std::set<size_t>   matched;          // detection indices already claimed

    // Step 2: greedy match — for each existing track pick the closest
    // unmatched detection within m_maxDistance. Outer loop is tracks
    // (matches PeopleTrackingAdapter.cpp lines 50-74).
    for (auto& kv : m_tracks) {
        auto& track = kv.second;

        int    bestIdx  = -1;
        float  bestDist = m_maxDistance;

        for (size_t i = 0; i < centroids.size(); ++i) {
            if (matched.count(i)) continue;
            const float dx = track.centroid.x - centroids[i].x;
            const float dy = track.centroid.y - centroids[i].y;
            const float d  = std::sqrt(dx*dx + dy*dy);
            if (d < bestDist) {
                bestDist = d;
                bestIdx  = static_cast<int>(i);
            }
        }

        if (bestIdx >= 0) {
            track.centroid    = centroids[bestIdx];
            track.missedFrames = 0;
            matched.insert(bestIdx);
            assignments[bestIdx] = track.id;
        }
    }

    // Step 3: every unmatched detection becomes a new track.
    for (size_t i = 0; i < centroids.size(); ++i) {
        if (matched.count(i)) continue;
        const int newId = m_nextId++;
        m_tracks[newId] = TrackedObject{newId, centroids[i], 0};
        assignments[static_cast<int>(i)] = newId;
    }

    // Step 4: prune stale tracks.
    for (auto it = m_tracks.begin(); it != m_tracks.end(); ) {
        if (it->second.missedFrames > m_maxMissed) it = m_tracks.erase(it);
        else                                       ++it;
    }

    return assignments;
}

std::map<int, int> CentroidTracker::update(
    const std::vector<cv::Point2f>& centroids,
    const std::vector<std::vector<float>>& embeddings,
    float appearanceWeight)
{
    for (auto& kv : m_tracks) kv.second.missedFrames++;

    std::map<int, int> assignments;
    std::set<size_t>   matched;
    const float        spatialW = 1.0f - appearanceWeight;

    // Greedy match per track. Cost = w_s * (d_spatial / maxDistance)
    //                              + w_a * (1 - cosine_sim).
    // Gating: the spatial term alone must remain below 1.0, i.e. raw distance
    // < m_maxDistance — a mostly-occluded detection can't teleport across the
    // frame just because its embedding matches.
    for (auto& kv : m_tracks) {
        auto& track = kv.second;

        int   bestIdx  = -1;
        float bestCost = std::numeric_limits<float>::max();

        for (size_t i = 0; i < centroids.size(); ++i) {
            if (matched.count(i)) continue;
            const float dx = track.centroid.x - centroids[i].x;
            const float dy = track.centroid.y - centroids[i].y;
            const float d  = std::sqrt(dx*dx + dy*dy);
            if (d >= m_maxDistance) continue;           // spatial gate

            const float sNorm = d / m_maxDistance;
            float cost;
            const bool hasEmb = i < embeddings.size() &&
                                !embeddings[i].empty() && !track.embedding.empty();
            if (hasEmb) {
                const float appDist = 1.0f - cosineSim(track.embedding, embeddings[i]);
                cost = spatialW * sNorm + appearanceWeight * appDist;
            } else {
                cost = sNorm;                            // spatial-only fallback
            }
            if (cost < bestCost) { bestCost = cost; bestIdx = static_cast<int>(i); }
        }

        if (bestIdx >= 0) {
            track.centroid     = centroids[bestIdx];
            track.missedFrames = 0;
            if (bestIdx < (int)embeddings.size())
                emaUpdate(track.embedding, embeddings[bestIdx]);
            matched.insert(bestIdx);
            assignments[bestIdx] = track.id;
        }
    }

    for (size_t i = 0; i < centroids.size(); ++i) {
        if (matched.count(i)) continue;
        const int newId = m_nextId++;
        TrackedObject t{newId, centroids[i], 0, {}};
        if (i < embeddings.size()) t.embedding = embeddings[i];
        m_tracks[newId] = std::move(t);
        assignments[static_cast<int>(i)] = newId;
    }

    for (auto it = m_tracks.begin(); it != m_tracks.end(); ) {
        if (it->second.missedFrames > m_maxMissed) it = m_tracks.erase(it);
        else                                       ++it;
    }

    return assignments;
}

File: python_examples/utils/tracker.py
"""
utils/tracker.py — Reusable greedy centroid tracker for DeepX demos.

Used by: people_tracking, smart_traffic, store_queue_analysis demos.
Algorithm: For each existing track, find the nearest unmatched detection
           within m_maxDistance. Unmatched detections become new tracks.
           Tracks missing for more than max_missing_frames are pruned.

This is a straight port of the greedy matching logic in
PeopleTrackingAdapter.cpp (lines 52-83). Coordinates are normalized
floats in [0.0, 1.0], so the distance threshold is frame-size agnostic
and the same tracker instance can be used across different resolutions.
"""

from __future__ import annotations

import math
from typing import Dict, List, Set, Tuple


class CentroidTracker:
    """Greedy nearest-neighbour centroid tracker.

    Each track holds a normalized (x, y) centroid, an age counter, and a
    running ``missing_frames`` count. On every ``update()`` call we:

    1. Bump ``missing_frames`` for every active track (we'll reset it
       below for any track that gets matched this frame).
    2. For each existing track, scan all detections and pick the closest
       one that hasn't been matched yet and is within ``max_distance``.
    3. Detections with no assigned track become brand-new tracks.
    4. Tracks that have been missing for longer than
       ``max_missing_frames`` are pruned.

    The tracker is deliberately stateless between frames apart from its
    internal ``_tracks`` dict — call ``reset()`` when the input source
    changes so stale IDs do not leak across videos.
    """

    def __init__(self, max_missing_frames: int = 10, max_distance: float = 0.1):
        """
        Args:
            max_missing_frames: How many frames a track can be missing before deletion.
            max_distance: Max normalized distance (0.0–1.0) for matching a detection to a track.
        """
        self._max_missing_frames = max_missing_frames
        self._max_distance = max_distance
        # Maps track_id -> {'id', 'centroid', 'missing_frames', 'age'}.
        self._tracks: Dict[int, Dict] = {}
        # Monotonic counter for new track IDs (matches C++ m_nextId).
        self._next_id: int = 0

    def update(
        self, centroids: List[Tuple[float, float]]
    ) -> Dict[int, int]:
        """
        Match current detections to existing tracks using greedy nearest-neighbor.

        Args:
            centroids: List of (x, y) normalized centroid positions from current frame.

        Returns:
            Dict mapping detection index (in input list) → assigned track ID.
            New detections get new IDs. Lost tracks are pruned internally.
        """
        # Step 1: every active track ages by one frame. Matched tracks will
        # have this decremented back to 0 below; unmatched ones keep the +1
        # which eventually trips the prune threshold.
        for track in self._tracks.values():
            track["missing_frames"] += 1
            track["age"] += 1

        # detection_index -> assigned_track_id
        assignments: Dict[int, int] = {}
        # Detections already claimed by a prior track this frame.
        matched_indices: Set[int] = set()

        # Step 2: outer loop is tracks (matches PeopleTrackingAdapter.cpp).
        # For each track, scan every detection and pick the closest unmatched
        # one within the distance threshold.
        for track_id, track in self._tracks.items():
            tx, ty = track["centroid"]

            best_idx = -1
            best_dist = self._max_distance  # Acts as both threshold and initial min.

            for i, (cx, cy) in enumerate(centroids):
                if i in matched_indices:
                    continue
                # Normalized Euclidean distance — math.sqrt keeps numpy optional.
                dist = math.sqrt((tx - cx) ** 2 + (ty - cy) ** 2)
                if dist < best_dist:
                    best_dist = dist
                    best_idx = i

            if best_idx >= 0:
                # Claim the detection, refresh the track's position, and
                # reset its missing-frame counter to 0.
                matched_indices.add(best_idx)
                assignments[best_idx] = track_id
                track["centroid"] = centroids[best_idx]
                track["missing_frames"] = 0

        # Step 3: unmatched detections become brand-new tracks.
        for i, centroid in enumerate(centroids):
            if i in matched_indices:
                continue
            new_id = self._next_id
            self._next_id += 1
            self._tracks[new_id] = {
                "id": new_id,
                "centroid": centroid,
                "missing_frames": 0,
                "age": 0,
            }
            assignments[i] = new_id

        # Step 4: prune stale tracks. We collect first, delete second, so
        # we do not mutate the dict while iterating over it.
        to_delete = [
            tid
            for tid, track in self._tracks.items()
            if track["missing_frames"] > self._max_missing_frames
        ]
        for tid in to_delete:
            del self._tracks[tid]

        return assignments

    def get_active_tracks(self) -> Dict[int, Dict]:
        """
        Returns all currently active tracks.
        Each track dict contains: 'id', 'centroid', 'missing_frames', 'age'
        """
        return self._tracks

    def reset(self):
        """Clear all tracks. Call when switching video source."""
        self._tracks.clear()
        self._next_id = 0
